From 9a2ed3340987cc65f20d967cce56255382eac829 Mon Sep 17 00:00:00 2001
From: Matthieu Gallien <matthieu_gallien@yahoo.fr>
Date: Tue, 19 Nov 2019 21:53:02 +0100
Subject: [PATCH 2/2] remove ki18n to avoid having it twice in v19.11.80

---
 CMakeLists.txt | 2 --
 1 file changed, 2 deletions(-)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index b3bbfe0..514bfa9 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -184,8 +184,6 @@ install(
     DESTINATION ${KDE_INSTALL_METAINFODIR}
 )
 
-ki18n_install(po)
-
 if (${ECM_VERSION} STRGREATER "5.58.0")
     install(FILES elisa.categories DESTINATION  ${KDE_INSTALL_LOGGINGCATEGORIESDIR})
 else()
-- 
2.24.0

